include/diy/master.hpp

```diff
@@ -861,211 +870,307 @@ void
+//              if (q not in memory)
+//                  load q
+//              issend(q)
+//
+//           test all requests
+//           if (iprobe)
+//               recv
+//           if (barrier_active)
+//               if (test barrier)
+//                   done = true
+//           else
+//               if (all sends finished and all queues have been processed (including out-of-core queues))
+//                   ibarrier
+//                   barrier_active = true
+// }
+//
+void
+diy::Master::
+rcomm_exchange(ToSendList& to_send, int out_queues_limit)
+{
+    IncomingRound   &current_incoming   = incoming_[exchange_round_];
+    bool            done                = false;
+    bool            ibarr_act           = false;
+    mpi::request    ibarr_req;                      // mpi request associated with ibarrier
 
-    OutgoingQueues& outgoing = outgoing_[from].queues;
-    for (OutgoingQueues::iterator it = outgoing.begin(); it != outgoing.end(); ++it)
+    while (!done)
     {
-      BlockID to_proc = it->first;
-      int     to      = to_proc.gid;
-      int     proc    = to_proc.proc;
-
-      log->debug("Processing queue:      {} <- {} of size {}", to, from, outgoing_[from].queues[to_proc].size());
+        send_outgoing_queues(to_send, out_queues_limit, current_incoming, true);
 
-      // There may be local outgoing queues that remained in memory
-      if (proc == comm_.rank())     // sending to ourselves: simply swap buffers
-      {
-        log->debug("Moving queue in-place: {} <- {}", to, from);
+        // kick requests
+        while (nudge());                            // TODO: use nudge, or loop over test instead? (nudge removes items from inflight_sends)
 
-        QueueRecord& in_qr  = current_incoming.map[to].records[from];
-        bool in_external  = block(lid(to)) == 0;
-        if (in_external)
+        check_incoming_queues(true);
+        if (ibarr_act)
         {
-          log->debug("Unloading outgoing directly as incoming: {} <- {}", to, from);
-          MemoryBuffer& bb = it->second;
-          in_qr.size = bb.size();
-          if (queue_policy_->unload_incoming(*this, from, to, in_qr.size))
-            in_qr.external = storage_->put(bb);
-          else
-          {
-            MemoryBuffer& in_bb = current_incoming.map[to].queues[from];
-            in_bb.swap(bb);
-            in_bb.reset();
-            in_qr.external = -1;
-          }
-        } else        // !in_external
+            if (ibarr_req.test())
+                done = true;
+        }
+        else
         {
-          log->debug("Swapping in memory:    {} <- {}", to, from);
-          MemoryBuffer& bb = current_incoming.map[to].queues[from];
-          bb.swap(it->second);
-          bb.reset();
-          in_qr.size = bb.size();
-          in_qr.external = -1;
+            // TODO: is the the right test for all queues have been processed (including out-of-core) ?
+            if (to_send.empty())
```

Comment:
This can and should be simplified to `if (to_send.empty() && inflight_sends_.empty())`. The inner tests are unnecessary and create some potential issues.